File: src/main/java/org/example/websitesalephone/dto/user/CreateUserDto.java
package org.example.websitesalephone.dto.user;

import lombok.Builder;
import lombok.Data;
import org.example.websitesalephone.entity.User;
import org.example.websitesalephone.utils.Constants;
import org.springframework.security.crypto.bcrypt.BCrypt;

import java.time.OffsetDateTime;

@Data
@Builder
public class CreateUserDto {

    private String id;

    private String loginId;

    private String fullName;

    private String profileImg;

    private String email;

    private String telNo;

    private String userCode;

    private String note;

    private String password;

    private RoleDto roleDto;

    public static User toEntity(CreateUserDto dto) {
        User user = new User();

        user.setId(dto.getId());
        user.setUsername(dto.getLoginId());
        user.setFullName(dto.getFullName() == null ? null : dto.getFullName().trim());
        user.setEmail(dto.getEmail());
        user.setDescription(dto.getNote());
        user.setPasswordHash(BCrypt.hashpw(dto.getPassword(), BCrypt.gensalt()));
        user.setPasswordExpiredAt(OffsetDateTime.now().plusDays(Constants.PASSWORD_EXPIRE_DAYS));
        user.setCodeUser(dto.getUserCode());
        user.setAvatar(dto.getProfileImg());

        return user;
    }
}

File: cms-user/src/models/CreateUserDto.ts
export interface IRoleDto {
    id?: string;
    name: string;
}

export interface ICreateUserDto {
    id?: string;
    loginId: string;
    fullName?: string;
    profileImg?: string;
    email?: string;
    telNo?: string;
    userCode?: string;
    note?: string;
    password: string;
    roleDto?: IRoleDto;
}

export class CreateUserDto {
    public id?: string;
    public loginId: string;
    public fullName?: string;
    public profileImg?: string;
    public email?: string;
    public telNo?: string;
    public userCode?: string;
    public note?: string;
    public password: string;
    public roleDto?: IRoleDto;

    constructor(data: {
        loginId: string;
        password: string;
        id?: string;
        fullName?: string;
        profileImg?: string;
        email?: string;
        telNo?: string;
        userCode?: string;
        note?: string;
        roleDto?: IRoleDto;
    }) {
        this.id = data.id?.trim();
        this.loginId = data.loginId.trim();
        this.fullName = data.fullName?.trim();
        this.profileImg = data.profileImg?.trim();
        this.email = data.email?.trim();
        this.telNo = data.telNo?.trim();
        this.userCode = data.userCode?.trim();
        this.note = data.note?.trim();
        this.password = data.password;
        this.roleDto = data.roleDto;
    }

    // Chuyển sang payload gửi API
    toPayload(): ICreateUserDto {
        return {
            id: this.id,
            loginId: this.loginId,
            fullName: this.fullName,
            profileImg: this.profileImg,
            email: this.email,
            telNo: this.telNo,
            userCode: this.userCode,
            note: this.note,
            password: this.password,
            roleDto: this.roleDto,
        };
    }
}
